text/3778-proc-macro-in-same-package-as-app.md

```diff
@@ -0,0 +1,95 @@
+3. In the main project, add the package as a dependency
+4. Implement the proc macro in the new package
+
+After this change, we create a new proc macro like this:
+1. Create a new directory called `proc-macro` alongside your `src` directory
+2. Implement the proc macro in a new file in `proc-macro`.
+
+To use the proc macro, simply import it via `crate::proc_macro`.
+```rust
+use crate::proc_macro::my_file::my_macro;
+```
+Or, if the file happens to be `mod.rs`, you can access it directly after the `proc_macro` bit.
+
+## Proc Macro Libraries
+Libraries like `syn`, `quote`, and `proc-macro2`, would be included under `[dev-dependecies]` in the cargo.toml. (Perhaps we should put it in build dependencies? or a new dependency section for proc macros.)
+
+## How it would work in the implementation
+Cargo would have to compile the `proc-macro` directory first, as a proc macro type (of course). Then, in compiling the main code, `crate::proc_macro::file_name::my_macro` would resolve the module to the file `proc-macro/file_name.rs`. Alternatively, if the user uses `mod.rs`, it would be resolved from `crate::proc_macro::my_macro`. This would finally be passed into rustc.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. The proc macro directory cannot use functions from src. (but that was not possible before anyways)
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+> Have proc macro files marked `#![proc_macro_file]` to signal to cargo to compile it first.
+
+Since it would compile first, proc macro files cannot import functions in the main code. The problem is having it side-by-side to the rest of your code makes it seem like you could just import it, when you cannot. Having it as a seperate directory makes clear of this.
+
+> Eliminate the need for new proc macro files/folders entirely, have the compiler work out where the proc macros are and separate them.
+
+This would suffer from the same issue as the last alternative, plus being harder to implement.
+
+> Introspection
+
+Harder to implement, with less payoff relative to the amount of work required. 
+
+# Prior art
+[prior-art]: #prior-art
+
+1. Zig comptime: metaprogramming code can sit directly next to application code.
+2. Declarative macros: can sit side by side as well, but is less powerful.
+3. Lisp macros: same as last two, except more powerful.
+4. `tests` directory, and `build.rs`: compiled at a different time as the main code.
+5. `Makefiles`, or other build systems: they allow for customisability for when code is built.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+1. Should proc macro dependencies be listed under `[dev-dependencies]`, `[build-dependencies]`, or a new `[proc-macro-dependencies]` section?
+2. Should we import like `crate::proc_macro::file::macro`, or via a new keyword, like `crate_macros::file::macro`? The latter would avoid name collisions, but might be more confusing.
```

Comment:
Without changing rustc itself or inserting some hidden prelude I don't see how can the proc-macro be available from `crate::proc_macro::*`.

Making it available from `crate_macros::*` is easy though, just add the flag `--extern crate_macros=/path/to/target/debug/deps/libxxxxx.dylib`.